images/hub/jupyterhub_config.py

```diff
@@ -46,46 +72,125 @@
+    ('lifecycle_hooks', None),
+    ('init_containers', None),
+    ('extra_containers', None),
+    ('mem_limit', 'memory.limit'),
+    ('mem_guarantee', 'memory.guarantee'),
+    ('cpu_limit', 'cpu.limit'),
+    ('cpu_guarantee', 'cpu.guarantee'),
+    ('extra_resource_limits', 'extraResource.limits'),
+    ('extra_resource_guarantees', 'extraResource.guarantees'),
+    ('environment', 'extraEnv'),
 ):
+    if cfg_key is None:
+        cfg_key = camelCaseify(trait)
     set_config_if_not_none(c.KubeSpawner, trait, 'singleuser.' + cfg_key)
-c.KubeSpawner.storage_extra_labels = get_config('singleuser.storage-extra-labels', {})
-
-c.KubeSpawner.image_spec = get_config('singleuser.image-spec')
-c.KubeSpawner.priority_class_name = get_config('singleuser.priority-class-name', "")
-
-c.KubeSpawner.tolerations.extend(get_config('singleuser.tolerations', []))
-c.KubeSpawner.node_selector.update(get_config('singleuser.node-selector', {}))
-c.KubeSpawner.node_affinity_required.extend(get_config('singleuser.node-affinity-required', []))
-c.KubeSpawner.node_affinity_preferred.extend(get_config('singleuser.node-affinity-preferred', []))
-c.KubeSpawner.pod_affinity_required.extend(get_config('singleuser.pod-affinity-required', []))
-c.KubeSpawner.pod_affinity_preferred.extend(get_config('singleuser.pod-affinity-preferred', []))
-c.KubeSpawner.pod_anti_affinity_required.extend(get_config('singleuser.pod-anti-affinity-required', []))
-c.KubeSpawner.pod_anti_affinity_preferred.extend(get_config('singleuser.pod-anti-affinity-preferred', []))
+
+image = get_config("singleuser.image.name")
+if image:
+    tag = get_config("singleuser.image.tag")
+    if tag:
+        image = "{}:{}".format(image, tag)
+
+    c.KubeSpawner.image_spec = image
+
+if get_config('singleuser.imagePullSecret.enabled'):
+    c.KubeSpawner.image_pull_secrets = 'singleuser-image-credentials'
+
+# scheduling:
+if get_config('scheduling.userScheduler.enabled'):
+    c.KubeSpawner.scheduler_name = os.environ['HELM_RELEASE_NAME'] + "-user-scheduler"
+if get_config('scheduling.podPriority.enabled'):
+    c.KubeSpawner.priority_class_name = os.environ['HELM_RELEASE_NAME'] + "-default-priority"
+
+# add node-purpose affinity
+match_node_purpose = get_config('scheduling.userPods.nodeAffinity.matchNodePurpose')
+if match_node_purpose:
+    node_selector = dict(
+        matchExpressions=[
+            dict(
+                key="hub.jupyter.org/node-purpose",
+                operator="In",
+                values=["user"],
+            )
+        ],
+    )
+    if match_node_purpose == 'prefer':
+        c.KubeSpawner.node_affinity_preferred.append(
+            dict(
+                weight=100,
+                preference=node_selector,
+            ),
+        )
+    elif match_node_purpose == 'require':
+        c.KubeSpawner.node_afinity_required.append(node_selector)
+    else:
+        raise ValueError("Unrecognized value for matchNodePurpose: %r" % match_node_purpose)
```

Comment:
I have documented `prefer`, `require` and `ignore`.

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/1f672e250ead51a8f43f2fc8d7cecf02e6bcdaaa/jupyterhub/schema.yaml#L706-L718

Reply:
Fixed.